gcc/rust/backend/rust-compile-expr.cc

```diff
@@ -211,13 +211,18 @@ CompileExpr::visit (HIR::MatchExpr &expr)
       return;
     }
 
-  rust_assert (scrutinee_expr_tyty->get_kind () == TyTy::TypeKind::ADT);
+  TyTy::TypeKind scrutinee_kind = scrutinee_expr_tyty->get_kind ();
+  rust_assert (scrutinee_kind == TyTy::TypeKind::BOOL
+	       || scrutinee_kind == TyTy::TypeKind::ADT);
 
-  // this will need to change but for now the first pass implementation, lets
-  // assert this is the case
-  TyTy::ADTType *adt = static_cast<TyTy::ADTType *> (scrutinee_expr_tyty);
-  rust_assert (adt->is_enum ());
-  rust_assert (adt->number_of_variants () > 0);
+  if (scrutinee_kind == TyTy::TypeKind::ADT)
+    {
+      // this will need to change but for now the first pass implementation,
+      // lets assert this is the case
+      TyTy::ADTType *adt = static_cast<TyTy::ADTType *> (scrutinee_expr_tyty);
+      rust_assert (adt->is_enum ());
+      rust_assert (adt->number_of_variants () > 0);
```

Comment:
The 3 assertions in there make me wonder if it wouldn't make sense to have a `rust_unimplemented` macro. We want to ICE in that case, but we also want the user to know that we just don't do proper error handling *yet* on match arms, or that we can't *yet* pattern match on structures, or can't *yet* produce a tasteful error for matchin on enums with 0 variants and so on.

I think it might make for a better user experience and encourage the users to report issues or contribute?

Reply:
I think the assertion makes sense here for now because we are saying that we only support bool and ADT enums right now. This patch made me realise I have been over thinking the case labels so actually this patch is really cool

Reply:
Oh definitely, it makes complete sense for now! I was pointing that out as a side issue :) Something we might wanna discuss and consider